core/app/models/spree/order.rb

```diff
@@ -870,6 +870,12 @@ def ensure_line_items_present
       end
     end
 
+    def ensure_positive_total
+      if outstanding_balance < 0
+        errors.add(:base, I18n.t('spree.cart_balance_negative')) && (return false)
```

Comment:
I don't think there's really any need to do a return of `false` here. The return value of this method will never be used and it's more for possibly adding errors to the record.

Reply:
Will happily remove it! Thanks for the feedback, appreciate it!

Reply:
Weird. I would suspect AASM to just check the errors. It's scary then that AASM lets me move forward in steps _with errors_ if the method is~, as Java would put it,~ `void`.

Reply:
the last method is returning !false hence the state machine continues. It's not raising an error mind you. It's just adding a model validation. if we were to `raise` an `Exception` here we would have stopped the state machine.

Reply:
@peterberkenbosch do you think raising an exception is a better alternative?

Reply:
I'm not sure, I think added the model validation presents it cleanly to the end user. When raising an exception, when not caught somewhere, we end up with 500 screens. 

On the other hand, when there is some sort of error tracking in place, we do send a quick and pressing message that something is not going as expected when we do raise an exception.

Reply:
Because it is possible for a user to hit this (given the right circumstances), I think we should leave it with AR. Thanks for clearing up our confusion!

Reply:
If raising an exception, would the state machine catch it and politely return `false` when doing `order.next` vs `order.next!`?

Reply:
`errors.add` is not returning `nil` it returns an Array with error messages